src/Credentials/CredentialProvider.php

```diff
@@ -267,32 +271,136 @@ public static function ecsCredentials(array $config = [])
     }
 
     /**
+     * Credential provider that creates credentials using assume role
+     * credentials from ini profile
+     *
+     * @param string|null $assume_role_profile assume_role_profile name to be used
+     *                                         By default: assumes_role for credentials file
+     *                                         profile assumes_role for config file.
+     * @param string|null $filename            If provided, uses a custom filename rather
+     *                                         than looking in the home directory.
+     * @param bool|false  $config_file         the type of file the assume_role_profile 
+     *                                         lives in. Use credentials file by default.
+     *
+     * @return callable
+     */
+    public static function assumeRole(
+        $assume_role_profile = null,
+        $filename = null,
+        $config_file = false
+    ){
+        $filename = self::getFileName($filename, $config_file);
+        $assume_role_profile = $assume_role_profile ?: (
+            $config_file ? 'profile assumes_role' : 'assumes_role'
+        );
+
+        return function () use ($assume_role_profile, $filename, $config_file) {
+            $data = self::checkProfileAvailability($assume_role_profile, $filename);
+            if (is_string($data)) {
+                return self::reject($data);
+            }
+            
+            if (!isset($data[$assume_role_profile]['role_arn'])) {
+                return self::reject("Profile specified is not an assume role profile.");
+            }
+
+            $args = self::loadConfigurationForAssumeRoleCredential(
+                $data,
+                $assume_role_profile,
+                $filename,
+                $config_file
+            );
+
+            if (!is_array($args)) {
+                // Rejected promise
+                return $args;
+            }
+            $provider = new AssumeRoleCredentialProvider($args);
+            return $provider();
+        };
+    }
+
+    public static function loadConfigurationForAssumeRoleCredential(
```

Comment:
I don't see a reason to publicly expose this. If you can't figure out a good way to test this if it's private, then let's add the `@internal` annotation to the docblock.